### PR TITLE
Reuse shared mem

### PR DESCRIPTION
**Pull request by psuriana**
Added logic for GPU to reuse shared memory. This tracks the liveness of each memory allocation and tries to coalesce together allocations which liveness does not overlap with each other. We prefer to coalesce dynamic-sized alloc with a dynamic-sized one and constant-sized alloc with a constant-sized one. If we can't find any free space which fulfills that criteria, we just coalesce with whichever space is more recently used. We prefer the most-recently freed space as stages that are close together usually have relatively similar allocation size. For constant-sized alloc, we do best-fit to find the candidate to coalesce. The free space can be smaller or larger than the alloc; we only care about the squared difference. It doesn't really matter since we take the max of the two as the new size.  

Example:

```
    f1(x, y) = x + y;
    f2(x, y) = f1(x+1, y) + f1(x, y+1);
    f3(x, y) = f2(x+1, y) + f2(x, y+1);
    f4(x, y) = f3(x+1, y) + f3(x, y+1);
```
```
    f4.compute_root().gpu_tile(x, y, 8, 8);
    f3.compute_at(f4, Var::gpu_blocks()).gpu_threads(x, y);
    f2.compute_at(f4, Var::gpu_blocks()).gpu_threads(x, y);
    f1.compute_at(f4, Var::gpu_blocks()).gpu_threads(x, y);
```
In the original code, this would use 1208 bytes of shared memory. The new logic will coalesce f1 and f3 and use 884 bytes of shared instead.

### CODE REVIEW

File: src/FuseGPUThreadLoops.cpp [hunk 1]
```diff
@@ -262,19 +266,52 @@ class ReplaceForWithIf : public IRMutator {
...
+    };
+
     struct SharedAllocation {
         string name;
         Type type;
         Expr size;
+        IntInterval liveness; // Start and end of the barrier stage at which this allocation is used.
+    };
+
+    struct AllocPairs {
```
Comment (abadams) [inline src/FuseGPUThreadLoops.cpp]: This looks more like a group of allocations than a pair.

File: src/FuseGPUThreadLoops.cpp [hunk 2]
```diff
@@ -293,6 +330,52 @@ class ExtractSharedAllocations : public IRMutator {
...
+            } else {
+                stmt = ProducerConsumer::make(op->name, produce, update, consume);
+            }
+        } else {
+            IRMutator::visit(op);
+        }
+    }
+
+    void visit(const Block *op) {
+        if (!in_threads && op->rest.defined()) {
```
Comment (abadams) [inline src/FuseGPUThreadLoops.cpp]: This looks like it's the same as the default visitor. Missing a barrier_stage++?

File: src/FuseGPUThreadLoops.cpp [hunk 3]
```diff
@@ -302,15 +385,15 @@ class ExtractSharedAllocations : public IRMutator {
...
+        shared.emplace(op->name, IntInterval(barrier_stage, barrier_stage));
         IRMutator::visit(op);
-        shared.erase(op->name);
         op = stmt.as<Allocate>();
         internal_assert(op);
 
         SharedAllocation alloc;
         alloc.name = op->name;
         alloc.type = op->type;
+        alloc.liveness = shared[op->name];
```
Comment (abadams) [inline src/FuseGPUThreadLoops.cpp]: It looks like shared[op->name].max gets updated below when you hit a load or store, but I don't see where alloc.liveness gets updated.

File: src/FuseGPUThreadLoops.cpp [hunk 4]
```diff
@@ -372,6 +458,115 @@ class ExtractSharedAllocations : public IRMutator {
...
+        // other type (e.g. pick constant-sized free space for a dynamic-sized allocation
+        // and vice versa). We prefer the most-recently freed space as stages that are
+        // close together usually have relatively similar allocation size. For
+        // constant-sized allocation, we prioritize free space which size differs
+        // the least with 'alloc' (can be smaller or larger; it does not really
+        // matter since we take the max of the two as the new size).
+
+        if (!is_const(alloc_size)) { // dynamic-sized alloc
+            for (int i = free_spaces.size() - 1; i >= 0; --i) {
+                // Only need to check the back of the vector since we always insert
```
Comment (abadams) [inline src/FuseGPUThreadLoops.cpp]: Is this comment still correct? It looks like you check the whole free list.

File: src/FuseGPUThreadLoops.cpp [hunk 5]
```diff
@@ -372,6 +458,115 @@ class ExtractSharedAllocations : public IRMutator {
...
+                } else if (free_idx == -1) {
+                    free_idx = i;
+                }
+            }
+        }
+
+        return free_idx;
+    }
+
+    // Given some allocations, return a vector of allocation pairs where each pair
```
Comment (abadams) [inline src/FuseGPUThreadLoops.cpp]: pair -> group

File: src/FuseGPUThreadLoops.cpp [hunk 6]
```diff
@@ -372,6 +458,115 @@ class ExtractSharedAllocations : public IRMutator {
...
+            for (int i = free_spaces.size() - 1; i >= 0; --i) {
+                // Only need to check the back of the vector since we always insert
+                // the most recent allocation at the back.
+                internal_assert(free_spaces[i] >= 0 && free_spaces[i] < (int)mem_allocs.size());
+                internal_assert(mem_allocs[free_spaces[i]].is_free(stage));
+
+                if (is_const(mem_allocs[free_spaces[i]].max_size_bytes)) {
+                    Expr size = alloc_size * alloc.type.bytes();
+                    Expr dist = mem_allocs[free_spaces[i]].max_size_bytes - size;
+                    const int64_t *current_diff = as_const_int(simplify(dist * dist));
```
Comment (abadams) [inline src/FuseGPUThreadLoops.cpp]: abs(dist) makes more sense to me than dist*dist (avoids having to think about overflow)

File: src/FuseGPUThreadLoops.cpp [hunk 7]
```diff
@@ -262,19 +266,54 @@ class ReplaceForWithIf : public IRMutator {
...
+    };
+
     struct SharedAllocation {
         string name;
         Type type;
         Expr size;
+        IntInterval liveness; // Start and end of the barrier stage at which this allocation is used.
+    };
+
+    struct AllocGroups {
```
Comment (dsharletg) [inline src/FuseGPUThreadLoops.cpp]: Should this be named `AllocGroup`? `AllocGroups` plural makes me think there is a set of groups, but I think this struct just represents one group of allocations.

File: src/FuseGPUThreadLoops.cpp [hunk 8]
```diff
@@ -372,6 +459,112 @@ class ExtractSharedAllocations : public IRMutator {
...
+                internal_assert(free_spaces[i] >= 0 && free_spaces[i] < (int)mem_allocs.size());
+                internal_assert(mem_allocs[free_spaces[i]].is_free(stage));
+
+                if (is_const(mem_allocs[free_spaces[i]].max_size_bytes)) {
+                    Expr size = alloc_size * alloc.type.bytes();
+                    Expr dist = mem_allocs[free_spaces[i]].max_size_bytes - size;
+                    const int64_t *current_diff = as_const_int(simplify(dist));
+                    internal_assert(current_diff != nullptr);
+                    int64_t abs_diff = std::abs(*current_diff);
+                    if ((free_idx == -1) || (abs_diff < diff)) {
```
Comment (dsharletg) [inline src/FuseGPUThreadLoops.cpp]: You could write this as:

    if ((free_idx == -1) || is_one(simplify(abs(dist) < diff))) {

And then you don't need any of the conversion to constants. I'm not sure if that's preferable or not :) It seems easier to read to me.

File: src/FuseGPUThreadLoops.cpp [hunk 9]
```diff
@@ -372,6 +459,112 @@ class ExtractSharedAllocations : public IRMutator {
...
+                    return true;
+                } else if (lhs.liveness.min == rhs.liveness.min) {
+                    return lhs.liveness.max < rhs.liveness.max;
+                }
+                return false;
+            }
+        );
+
+        vector<AllocGroup> mem_allocs;
+        vector<int> free_spaces; // Contains index to free spaces in mem_allocs
```
Comment (dsharletg) [inline src/FuseGPUThreadLoops.cpp]: Would it be easier to have a 'bool free' member of AllocGroup?

**Pull request merged**